### ISSUE TITLE
Support feature selection

### ISSUE THREAD

**Issue by amaya382** [myui/hivemall#338]
# Feature selection

Feature selection is the process of selecting a subset consisting of influential features from multiple features. It is an important technique to **enhance results**, **shorten training time** and **make features human-understandable**.

*Currently, following is temporary I/F*

#### Input

| array<features::string> input | int num_of_features | [string method = chi2] |
|:--:|:--:|:--:|
| feature vector | number of features to be selected | method to select features |

#### Output

return only one row

| array<string> features |
|:--:|
| selected features' name |

#### Input

| array\<features::string> features | int num_of_features | [string method = chi2] |
|:--:|:--:|:--:|
| feature vector | number of features to be selected | method to select features |

#### Output

| array\<features::string> features |
|:--:|
| selected features |

#### Output

# To discuss

* how to receive features
* which is better, creating UDF specialized in each mehods or generic one, which branch by literal option
* return selected features' name, features themselves or both

**Comment by myui** [comment]
@amaya382 depends on input format of each algorithm. If we can use same format as the input, better to provide a `feature_selection` function w/ an algorithm option.

Also, how to apply distributed processing is another issue. mRMR is processed in parallel by MapReduce in some implementation.

**Comment by myui** [comment]
@amaya382 we can do better as follows:

```
X::relation(array<double>)  # n_features * n_examples matrix
Y::relation(array<int>)     # n_class * n_examples matrix

observed = dotp(Y.T, X)     # n_classes * n_features matrix
UDAF transpose_and_dot(Y, X)::array<array<double>>

feature_count = X.sum       # n_features col vector
UDAF array_sum(X)::array<double>

class_prob = Y.mean         # n_class col vector
UDAF array_avg(Y)::array<double>

expected = dotp(class_prob.T, feature_count)    # n_class * n_features matrix
UDAF transpose_and_dot(class_prob, feature_count)::array<array<double>>

chi2,pval = chisquare(observed, expected)   # n_class * n_features => n_features
UDAF chi2(observerd::array<double>, expected::array<double>)::struct<array<double>,array<double>>

chi2    # n_features col vector
pval    # n_features col vector
```

```sql
create table input (
  X array<double>,
  Y array<int>
);
WITH stats (
  select
    transpose_and_dot(Y, X) as observed, -- array<array<double>> # n_classes * n_features matrix
    array_sum(X) as feature_count, -- n_features col vector # array<double>
    array_avg(Y) as class_prob -- n_class col vector # array<double>
  from
    input
),
test as (
  select
    observed, 
    transpose_and_dot(class_prob, observed) as expected -- array<array<double>> # n_class * n_features matrix
  from
    stats
),
select
  chi2(observed, expected) as (chi2, pval) -- struct<array<double>,array<double>>
from
  test
```

**Comment by amaya382** [comment]
@myui okay, I'll try this strategy

**Comment by myui** [comment]
@amaya382 could you validate your implementation to other chi2 implementation (e.g., scikit-learn) in the unit test.

**Comment by amaya382** [comment]
@myui 

## Iris dataset
### results by hivemall with systemtest, which exec query actually
|  | f0 | f1 | f2 | f3 |
|--:|:--:|:--:|:--:|:--:|
| **chi2_vals** | 10.817820878493995 | 3.5944990176817315 | 116.16984746363957 | 67.24482558215503 |
| **p_vals** | 0.004476514990225833 | 0.16575416718561453 | 0.0 | 2.55351295663786E-15 |

### results by scikit-learn
|  | f0 | f1 | f2 | f3 |
|--:|:--:|:--:|:--:|:--:|
| **chi2_vals** | 10.81782088 | 3.59449902 | 116.16984746 | 67.24482759 |
| **p_vals** | 4.47651499e-03 | 1.65754167e-01 | 5.94344354e-26 | 2.50017968e-15 |

**Comment by amaya382** [comment]
@myui 

## SNR (Iris dataset)
### results by hivemall with systemtest, which exec query actually, incremental algorithm
|  | f0 | f1 | f2 | f3 |
|--:|:--:|:--:|:--:|:--:|
| **aggregated SNR** | 3.2700426804423826 | 1.9034599450433007 | 11.359577906085278 | 9.794847727147216 |

### results by python-numpy, batch algorithm
|  | f0 | f1 | f2 | f3 |
|--:|:--:|:--:|:--:|:--:|
| **aggregated SNR** | 3.27004268 | 1.90345995 | 11.35957791 | 9.79484773 |

Also already tested on EMR, worked properly
